osquery/ev2/buffered_subscription.h

```diff
@@ -0,0 +1,153 @@
+#include <string>
+
+namespace osquery {
+namespace ev2 {
+
+/**
+ * @brief Simple queue based ev2::Subscription.
+ *
+ * @details Provides a basic queue based ev2::Subscription buffering events for
+ * the subscriber.
+ */
+template <typename EventT>
+class BufferedSubscription : public Subscription {
+ public:
+  /**
+   * @brief BufferedSubscription constructor.
+   *
+   * @params subscriber See ev2::Subscription::Subscription()
+   * @params pub_type See ev2::Subscription::Subscription()
+   */
+  explicit BufferedSubscription(std::string subscriber,
+                                std::type_index pub_type)
+      : Subscription(std::move(subscriber), std::move(pub_type)),
+        abort_(false),
+        waiting_(0) {}
+  virtual ~BufferedSubscription() = default;
+
+  /**
+   * @brief See ev2::Subscription::avail().
+   *
+   * @details Returns the current total number of events buffered. As expected
+   * the value returned by this method increases after calling enqueue() and
+   * decreases after calling take() and won't be change otherwise.
+   */
+  std::size_t avail() const override {
+    std::unique_lock<std::mutex> lock(buffer_mutex_);
+
+    return buffer_.size();
+  }
+
+  /**
+   * @brief See ev2::Subscription::wait().
+   *
+   * @details This method is thread-safe as producer and consumer will probably
+   * run on different threads, however it is not designed to handle
+   * multiple simultaneous consumers. In case multiple threads call wait
+   * simultaneously the implementation can only guarantee that
+   * one of the pending threads will be released once the batch size is reached
+   * however other threads might as well be released. The wait timeout will
+   * always be respected. If you still want to use multiple consumer threads it
+   * is up to you to decide how to handle the consumption of events. Remember
+   * that take() shouldn't be called without guaranteeing first that there are
+   * pending events by either calling wait() or avail().
+   */
+  std::size_t wait(std::size_t batch = 1,
+                   std::chrono::milliseconds timeout =
+                       std::chrono::milliseconds::zero()) override {
+    waiting_++;
+
+    std::unique_lock<std::mutex> lock(buffer_mutex_);
+
+    if (timeout == std::chrono::milliseconds::zero()) {
+      buffer_cv_.wait(
+          lock, [this, batch] { return buffer_.size() >= batch || abort_; });
+    } else {
+      buffer_cv_.wait_for(lock, timeout, [this, batch] {
+        return buffer_.size() >= batch || abort_;
+      });
+    }
+
+    waiting_--;
+
+    return buffer_.size();
+  }
+
+  /**
+   * @brief See ev2::Subscription::abort()
+   *
+   * @details Calling abort() will block until all threads blocked on wait()
+   * exit. Calling abort() will leave the object in a working state and it is
+   * safe to call to call any method on the object as soon as abort() exits. No
+   * events will be lost in by calling abort().
+   */
+  void abort() override {
```

Comment:
hi @fmanco ,  I believe abort_ should be set under the protection of the lock else there is a race.
{
  std::unique_lock<std::mutex> lock(buffer_mutex_);
  abort_ = true;
  buffer_cv_.notify_all(); // keep this inside or outside of the scope  your choice
}

Reply:
What is the race condition you're seeing?

Reply:
My reasoning is simple, abort_ is shared by two (or threads) so it should be accessed under the protection of a lock. 

I can give you an example where it can cause 'while (waiting_ != 0) {}' to spin for indefinite period of time.
At present abort_ is being used as relaxed-atomic  (atomic since on x64 aligned object are atomic in nature).

Reply:
Yeah, `abort_` is being used as an atomic. Just out of curiosity can you give the example where the abort loop would spin forever?

Reply:
Suppose there are two threads

    Time line (T0<T1<T2)
           thread 1                               |            thread2
    T0    abort_ = true;                      |
    T1   buffer_cv_.notify_all();          | 
    T2                                               |   buffer_cv_.wait(
                                                               lock, [this, batch] { return                                        buffer_.size() >= batch || 
                                                  abort_; });          

In the above example if  condition (buffer_.size() >= batch) is false and decision is to be taken based upon the (abort_==true) then do not assume that abort_ is true since it happened at T0 and it is being checked at time T2.  Since abort_ can be false so thread2 can go back to sleep on condition variable. Since waiting_ has value >0 which will cause thread1 to loop.